<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_21) on Thu Sep 19 21:31:26 EDT 2013 -->
<title>FeatureBuilder</title>
<meta name="date" content="2013-09-19">
<link rel="stylesheet" type="text/css" href="../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="FeatureBuilder";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/FeatureBuilder.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev Class</li>
<li>Next Class</li>
</ul>
<ul class="navList">
<li><a href="../index.html?learningBasedPlacer/FeatureBuilder.html" target="_top">Frames</a></li>
<li><a href="FeatureBuilder.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field_summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field_detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">learningBasedPlacer</div>
<h2 title="Class FeatureBuilder" class="title">Class FeatureBuilder</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li>learningBasedPlacer.FeatureBuilder</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<hr>
<br>
<pre>public class <span class="strong">FeatureBuilder</span>
extends java.lang.Object</pre>
<div class="block">Converts binary output serialized processed protein files into libsvm input files.</div>
<dl><dt><span class="strong">Version:</span></dt>
  <dd>1.0</dd>
<dt><span class="strong">Author:</span></dt>
  <dd>Steve Morse</dd></dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field_summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#CHLOROPLAST">CHLOROPLAST</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#ENDOPLASMIC_RETICULUM">ENDOPLASMIC_RETICULUM</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#MITOCHONDRION">MITOCHONDRION</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#NUCLEUS">NUCLEUS</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#OTHER">OTHER</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#PEROXISOME">PEROXISOME</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static int</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#SECRETORY_PATHWAY">SECRETORY_PATHWAY</a></strong></code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#FeatureBuilder()">FeatureBuilder</a></strong>()</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method_summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span>Methods</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#build()">build</a></strong>()</code>
<div class="block">Method that does the controls the conversion of protein output files into libsvm
 input text files, a index integer for a location and a series of label:value pairs for
 features.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#getLocationValue(protein.Protein)">getLocationValue</a></strong>(<a href="../protein/Protein.html" title="class in protein">Protein</a>&nbsp;currentProtein)</code>
<div class="block">Assigns a location value integer code to a protein based on text or GO code
 placement data from the protein matched against the subset GO entries for each
 supported cell location.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#incrementDiAminoValue(java.lang.String, int, java.util.Map, java.util.List)">incrementDiAminoValue</a></strong>(java.lang.String&nbsp;diAmino,
                     int&nbsp;lengthOfSequence,
                     java.util.Map&lt;java.lang.String,java.lang.Integer&gt;&nbsp;valueIndicesMap,
                     java.util.List&lt;java.lang.Float&gt;&nbsp;featureValuesForOneEntry)</code>
<div class="block">increments a di-amino (a amino acid bigram) ratio of total protein amino acid's value</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#incrementSingleAminoValue(java.lang.Character, int, java.util.Map, java.util.List)">incrementSingleAminoValue</a></strong>(java.lang.Character&nbsp;c,
                         int&nbsp;lengthOfSequence,
                         java.util.Map&lt;java.lang.String,java.lang.Integer&gt;&nbsp;valueIndicesMap,
                         java.util.List&lt;java.lang.Float&gt;&nbsp;featureValuesForOneEntry)</code>
<div class="block">increments a single amino acid's ratio of total protein amino acid's value</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#loadValueIndicesMap(java.util.Map)">loadValueIndicesMap</a></strong>(java.util.Map&lt;java.lang.String,java.lang.Integer&gt;&nbsp;valueIndicesMap)</code>
<div class="block">loads the valueIndicesMap with feature names and indices</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.util.List&lt;java.lang.Float&gt;</code></td>
<td class="colLast"><code><strong><a href="../learningBasedPlacer/FeatureBuilder.html#makeNewvaluesForOneFeatureList(int)">makeNewvaluesForOneFeatureList</a></strong>(int&nbsp;sizeOfList)</code>
<div class="block">Makes a new empty (zero filled) feature value list for an data entry</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>equals, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field_detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="CHLOROPLAST">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>CHLOROPLAST</h4>
<pre>public static final&nbsp;int CHLOROPLAST</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../constant-values.html#learningBasedPlacer.FeatureBuilder.CHLOROPLAST">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="ENDOPLASMIC_RETICULUM">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>ENDOPLASMIC_RETICULUM</h4>
<pre>public static final&nbsp;int ENDOPLASMIC_RETICULUM</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../constant-values.html#learningBasedPlacer.FeatureBuilder.ENDOPLASMIC_RETICULUM">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="MITOCHONDRION">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>MITOCHONDRION</h4>
<pre>public static final&nbsp;int MITOCHONDRION</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../constant-values.html#learningBasedPlacer.FeatureBuilder.MITOCHONDRION">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="NUCLEUS">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>NUCLEUS</h4>
<pre>public static final&nbsp;int NUCLEUS</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../constant-values.html#learningBasedPlacer.FeatureBuilder.NUCLEUS">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="PEROXISOME">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>PEROXISOME</h4>
<pre>public static final&nbsp;int PEROXISOME</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../constant-values.html#learningBasedPlacer.FeatureBuilder.PEROXISOME">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="SECRETORY_PATHWAY">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>SECRETORY_PATHWAY</h4>
<pre>public static final&nbsp;int SECRETORY_PATHWAY</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../constant-values.html#learningBasedPlacer.FeatureBuilder.SECRETORY_PATHWAY">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="OTHER">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>OTHER</h4>
<pre>public static final&nbsp;int OTHER</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../constant-values.html#learningBasedPlacer.FeatureBuilder.OTHER">Constant Field Values</a></dd></dl>
</li>
</ul>
</li>
</ul>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="FeatureBuilder()">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>FeatureBuilder</h4>
<pre>public&nbsp;FeatureBuilder()</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method_detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="build()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>build</h4>
<pre>public&nbsp;void&nbsp;build()</pre>
<div class="block">Method that does the controls the conversion of protein output files into libsvm
 input text files, a index integer for a location and a series of label:value pairs for
 features.</div>
</li>
</ul>
<a name="getLocationValue(protein.Protein)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getLocationValue</h4>
<pre>public&nbsp;int&nbsp;getLocationValue(<a href="../protein/Protein.html" title="class in protein">Protein</a>&nbsp;currentProtein)</pre>
<div class="block">Assigns a location value integer code to a protein based on text or GO code
 placement data from the protein matched against the subset GO entries for each
 supported cell location.</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>currentProtein</code> - the protein to have a location value assessed</dd>
<dt><span class="strong">Returns:</span></dt><dd>the location value integer code for that protein</dd></dl>
</li>
</ul>
<a name="incrementSingleAminoValue(java.lang.Character, int, java.util.Map, java.util.List)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>incrementSingleAminoValue</h4>
<pre>public&nbsp;void&nbsp;incrementSingleAminoValue(java.lang.Character&nbsp;c,
                             int&nbsp;lengthOfSequence,
                             java.util.Map&lt;java.lang.String,java.lang.Integer&gt;&nbsp;valueIndicesMap,
                             java.util.List&lt;java.lang.Float&gt;&nbsp;featureValuesForOneEntry)</pre>
<div class="block">increments a single amino acid's ratio of total protein amino acid's value</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>c</code> - a single amino acid fasta code</dd><dd><code>lengthOfSequence</code> - the total length of the protein sequence</dd><dd><code>valueIndicesMap</code> - A Map associating feature names with their indices in 
 the featureValuesForOneEntry List</dd><dd><code>featureValuesForOneEntry</code> - a list of all the feature values for one input entry</dd></dl>
</li>
</ul>
<a name="incrementDiAminoValue(java.lang.String, int, java.util.Map, java.util.List)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>incrementDiAminoValue</h4>
<pre>public&nbsp;void&nbsp;incrementDiAminoValue(java.lang.String&nbsp;diAmino,
                         int&nbsp;lengthOfSequence,
                         java.util.Map&lt;java.lang.String,java.lang.Integer&gt;&nbsp;valueIndicesMap,
                         java.util.List&lt;java.lang.Float&gt;&nbsp;featureValuesForOneEntry)</pre>
<div class="block">increments a di-amino (a amino acid bigram) ratio of total protein amino acid's value</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>diAmino</code> - the bigram to increment the value of</dd><dd><code>lengthOfSequence</code> - the total length of the protein sequence</dd><dd><code>valueIndicesMap</code> - A Map associating feature names with their indices in 
 the featureValuesForOneEntry List</dd><dd><code>featureValuesForOneEntry</code> - a list of all the feature values for one input entry</dd></dl>
</li>
</ul>
<a name="loadValueIndicesMap(java.util.Map)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>loadValueIndicesMap</h4>
<pre>public&nbsp;void&nbsp;loadValueIndicesMap(java.util.Map&lt;java.lang.String,java.lang.Integer&gt;&nbsp;valueIndicesMap)</pre>
<div class="block">loads the valueIndicesMap with feature names and indices</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>valueIndicesMap</code> - the Map to be loaded with name and index values</dd></dl>
</li>
</ul>
<a name="makeNewvaluesForOneFeatureList(int)">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>makeNewvaluesForOneFeatureList</h4>
<pre>public&nbsp;java.util.List&lt;java.lang.Float&gt;&nbsp;makeNewvaluesForOneFeatureList(int&nbsp;sizeOfList)</pre>
<div class="block">Makes a new empty (zero filled) feature value list for an data entry</div>
<dl><dt><span class="strong">Parameters:</span></dt><dd><code>sizeOfList</code> - the size of the list to make and zero fill</dd>
<dt><span class="strong">Returns:</span></dt><dd>the newly minted list</dd></dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="class-use/FeatureBuilder.html">Use</a></li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../deprecated-list.html">Deprecated</a></li>
<li><a href="../index-files/index-1.html">Index</a></li>
<li><a href="../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li>Prev Class</li>
<li>Next Class</li>
</ul>
<ul class="navList">
<li><a href="../index.html?learningBasedPlacer/FeatureBuilder.html" target="_top">Frames</a></li>
<li><a href="FeatureBuilder.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field_summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field_detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
